src/ray/gcs/store_client/redis_store_client.cc

```diff
@@ -386,6 +386,7 @@ Status RedisStoreClient::RedisScanner::ScanKeys(
 
 void RedisStoreClient::RedisScanner::Scan(std::string match_pattern,
                                           const StatusCallback &callback) {
+  absl::MutexLock lock(&mutex_);
```

Comment:
technically it can guard the for loop that comes later. and the issue is probably somehow the iterator was invalidated by `OnScanCallback`'s `shard_to_cursor_ .erase`.

Reply:
![image](https://user-images.githubusercontent.com/13081808/108648717-e48aa380-74f6-11eb-8e37-909347b159b4.png)
If `io_services` is not `main_io_service`, there will be multithreading problem. If it's `main_io_service`, there should be no problem in theory.